engine/server/engine/server/websocket_api_handler.go

```diff
@@ -0,0 +1,386 @@
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_interface/objects/enclave"
+	user_service "github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_interface/objects/service"
+	"github.com/kurtosis-tech/kurtosis/engine/server/engine/centralized_logs"
+	"github.com/kurtosis-tech/kurtosis/engine/server/engine/centralized_logs/client_implementations/persistent_volume/log_file_manager"
+	"github.com/kurtosis-tech/kurtosis/engine/server/engine/centralized_logs/logline"
+	"github.com/kurtosis-tech/kurtosis/engine/server/engine/enclave_manager"
+	"github.com/kurtosis-tech/kurtosis/engine/server/engine/utils"
+	"github.com/kurtosis-tech/kurtosis/metrics-library/golang/lib/metrics_client"
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/labstack/echo/v4"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"golang.org/x/net/websocket"
+
+	api_type "github.com/kurtosis-tech/kurtosis/api/golang/http_rest/api_types"
+	api "github.com/kurtosis-tech/kurtosis/api/golang/http_rest/websocket_api"
+)
+
+type WebSocketRuntime struct {
+	// The version tag of the engine server image, so it can report its own version
+	ImageVersionTag string
+
+	EnclaveManager *enclave_manager.EnclaveManager
+
+	// The protected user ID for metrics analytics purpose
+	MetricsUserID string
+
+	// User consent to send metrics
+	DidUserAcceptSendingMetrics bool
+
+	// The clients for consuming container logs from the logs' database server
+
+	// per week pulls logs from enclaves created post log retention feature
+	PerWeekLogsDatabaseClient centralized_logs.LogsDatabaseClient
+
+	// per file pulls logs from enclaves created pre log retention feature
+	// TODO: remove once users are fully migrated to log retention/new log schema
+	PerFileLogsDatabaseClient centralized_logs.LogsDatabaseClient
+
+	LogFileManager *log_file_manager.LogFileManager
+
+	MetricsClient metrics_client.MetricsClient
+}
+
+func sendErrorCode(ctx echo.Context, code int, message string) error {
+	err := ctx.JSON(code, message)
+	return err
+}
+
+type LogStreamer struct {
+	serviceLogsByServiceUuidChan chan map[user_service.ServiceUUID][]logline.LogLine
+	errChan                      chan error
+	cancelCtxFunc                func()
+	requestedServiceUuids        []user_service.ServiceUUID
+	notFoundServiceUuids         []string
+}
+
+func (engine WebSocketRuntime) GetEnclavesEnclaveIdentifierLogs(ctx echo.Context, enclaveIdentifier api_type.EnclaveIdentifier, params api.GetEnclavesEnclaveIdentifierLogsParams) error {
+	streamer, err := engine.getLogStreamer(
+		ctx,
+		enclaveIdentifier,
+		utils.MapList(params.ServiceUuidSet, func(x string) user_service.ServiceUUID { return user_service.ServiceUUID(x) }),
+		params.FollowLogs,
+		params.ReturnAllLogs,
+		utils.MapPointer(params.NumLogLines, func(x int) uint32 { return uint32(x) }),
+		params.ConjunctiveFilters,
+	)
+	if err != nil {
+		logrus.Error(err)
```

Comment:
I think we could add more context in this error, "Failed to get log streamer with error:" for instance